File: ruby/025-word-count/word_count.rb
class Phrase
  attr_reader :sentence

  def initialize(sentence)
    @sentence = sentence
  end

  def word_count
    result = {}
    sentence.split(/[^'\w]/).each do |word|
      downcase_word = word.downcase
      downcase_word = downcase_word[1..] if downcase_word[0] == "'"
      downcase_word = downcase_word[..-2] if downcase_word[-1] == "'"
      next if ['', "'"].include?(downcase_word)


      result[downcase_word] ? result[downcase_word] += 1 : result[downcase_word] = 1
    end
    result
  end
end

File: go/060-word-count/word_count.go
package wordcount

import (
	"regexp"
	"strings"
)

type Frequency map[string]int

func WordCount(phrase string) Frequency {
	delimiter := regexp.MustCompile(`[^a-zA-Z0-9']`)
	wordArray := delimiter.Split(phrase, -1)
	freq := make(Frequency)

	for _, word := range wordArray {
		word := strings.Trim(word, "'")
		if len(word) == 0 {
			continue
		}

		freq[strings.ToLower(word)]++
	}

	return freq
}

// Community Solution
//func WordCount(sentence string) Frequency {
//	result := make(Frequency)
//	reg := regexp.MustCompile(`\w+('\w+)?`)
//	for _, word := range reg.FindAllString(strings.ToLower(sentence), -1) {
//		result[word]++
//	}
//	return result
//}
